pkg/util/slices_test.go

```diff
@@ -79,3 +79,67 @@ func TestCombinesUniqueStrings(t *testing.T) {
 		require.Equal(t, tc.expected, output)
 	}
 }
+
+func TestFilter(t *testing.T) {
+	numbers := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
+	evens := Filter(numbers, func(i int) bool {
+		return i%2 == 0
+	})
+	odds := Filter(numbers, func(i int) bool {
+		return i%2 == 1
+	})
+
+	// Assert filtering works.
+	require.Equal(t, []int{1, 3, 5, 7, 9}, odds)
+	require.Equal(t, []int{0, 2, 4, 6, 8}, evens)
+
+	// Assert/demonstrate that filtering does not mutate the original slice.
+	require.Equal(t, []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}, numbers)
+
+	// The filtered slices' capacity is correctly set and the original slice's
+	// capacity is unmodified.
+	require.Equal(t, len(odds), cap(odds))
+	require.Equal(t, len(evens), cap(evens))
+	require.Equal(t, len(numbers), cap(numbers))
+
+	// And some weird cases.
+	require.Equal(t, []int{}, Filter(nil, func(i int) bool {
+		return true
+	}))
+	require.Equal(t, []int{}, Filter([]int{}, func(i int) bool {
+		return true
+	}))
+	require.Equal(t, []int{}, Filter(numbers, func(i int) bool {
+		return false
+	}))
+}
+
+func TestMap(t *testing.T) {
+	require.Equal(t, []bool{}, Map(nil, func(i int) bool {
+		return true
+	}))
+	require.Equal(t, []bool{}, Map([]int{}, func(i int) bool {
+		return true
+	}))
+	require.Equal(t, []bool{false, true, false, true, false}, Map([]int{1, 2, 3, 4, 5}, func(i int) bool {
+		return i%2 == 0
+	}))
+}
+
+func TestMapBy(t *testing.T) {
+	require.Equal(t, map[int]bool{}, MapBy(nil, func(i int) (int, bool) {
+		return 0, false
```

Comment:
Nit: adding a `t.Fail()` in this function would help guarantee that this function is not called if the collection is empty, an important invariant.

Reply:
Done!